Invitation workflow for OCM, take 2

glpatcern opened the pull request:
This PR supersedes https://github.com/cs3org/OCM-API/pull/41.

I have rebased the new endpoints on top of the current `develop` branch. We should have a discussion with @michielbdejong as a main contributor in this repo also to address the yet outstanding comments from the previous PR, so this is still WIP though Reva currently includes a full implementation of these specs.

LovisaLugnegard committed:
Add endpoints for accept and forward invite

labkode committed:
ocm: enhance invitation flow

labkode committed:
ocm: fix

labkode committed:
fix leading slash

LovisaLugnegard committed:
Smaller copy updates

LovisaLugnegard committed:
Minor updates in the invites API

LovisaLugnegard committed:
Fix typos

Commit:
Add name and email to invites/accept request

LovisaLugnegard committed:
Improve description

glpatcern committed:
Removed further conflicts

michielbdejong commented:
@glpatcern great, let's have a video call to walk through the details!

glpatcern committed:
Renamed `accept` -> `accepted`, following @michielbdejong suggestions

glpatcern committed:
Extended the description of a share `resourceType` to cover web apps

glpatcern reviewed (commented):
(Self-reviewing for further brainstorming)

glpatcern commented on spec.yaml:
```diff
@@ -259,7 +295,9 @@ definitions:
-          Resource type (file, calendar, contact,...)
+          Resource type (file, webapp, calendar, contact, ...). A `webapp` share
+          is an application that can be opened on the remote end, typically
+          to enable collaborative workflows.
```
This is a (non-binding) idea to further diversify the type of resource one is sharing, including the application engine to use to edit some files.

However, the action of opening a file in a web app would come _after_ the file was already shared with an `/ocm/share` call (and `resourceType=file`), so I'm not entirely sure an additional `/ocm/share` call with `resourceType=webapp` is the right semantic here.

That said, if the two EFSS both use Reva as the reference implementation of OCM, then the `open-in-app` workflow implemented in Reva ought to actually work out of the box, provided that an option (e.g. in the context menu) is given to the user in the web frontend to execute a "remote open" as opposed to a "local open". That hook would be served by the local Reva, which will forward it to the remote Reva as it happens on an OCM share: and the remote Reva will ultimately perform the "remote open", returning the app URL.

glpatcern replied:
Resuming the discussion and targeting a resolution at the upcoming workshop @ CERN:

Even assuming that all EFSSs use Reva as the reference implementation of OCM (which is the target model in ScienceMesh), it would still be good to expose an `/ocm/open` endpoint, which Reva would simply forward to `/app/open`. In this context, a `webapp` resource type is indeed redundant, but on the OCM share part some `edit in app` kind of permission may be envisaged to enable such workflow, where the "normal" r/w permission would simply allow to download/upload data to the share.

glpatcern commented on spec.yaml:
```diff
@@ -140,15 +137,54 @@ paths:
               type: string
           schema:
             $ref: '#/definitions/Error'
+  /invites/forward:
```
For the record, there was also the idea to rename `/forward` as well, but I'm not sure something like `/deliver` would be much better.

On the other hand, I support renaming `/accept` to `/accepted` as this is really a message sent back to inform that the invitation was accepted.

glpatcern replied:
Discussing with @michielbdejong, the `forward` action is actually called `accept` in the OC/NC apps, and there's no explicit OCM `accept` action in those apps.

smesterheide commented:
I can't speak to the issue at hand, but I am happy to see progress in advancing the spec to a new version. Is there a process to follow?

glpatcern commented:
> I can't speak to the issue at hand, but I am happy to see progress in advancing the spec to a new version. Is there a process to follow?

Hi @smesterheide, if you have some specific suggestion or RFE you are welcome to file an issue or a PR in GitHub. More in general the [CS3 community](https://www.cs3community.org/) regularly meets and discusses OCM issues among other topics, and I expect we'll tag a new version of the API at next event.

glpatcern committed:
Reverted previous commit, a webapp resource type is actually redundant

glpatcern commented:
@michielbdejong @labkode I propose to get this fully sorted out and merged in these days. At least it would reflect what is already (partially) implemented in Reva and in the OC10/NC integrations, as far as I understand.

glpatcern committed:
Refinements following discussions with @michielbdejong and @labkode:

- The /forward endpoint was removed as it's a local call
- The endpoint to notify about an accepted invite is now /invite-accepted
  and it is the only additional endpoint in this PR.
- The response of that endpoint now contains the identity details
  of the sender, which are safe to be disclosed at this point as
  the receiver already did. This allows to address issue #55.
- A new protocol `webapp` is added to `/NewShare`, to enable opening
  files with remote apps. This removes the need of yet another
  dedicated endpoint.

glpatcern commented:
@michielbdejong and @labkode I've tried to spell out what we brainstormed yesterday (and thanks for the nice discussion!). Please have a look so we can move forward with the necessary changes in Reva.

glpatcern committed:
In /NewShare, protocol should be an array

glpatcern committed:
Formatting changes

michielbdejong commented on spec.yaml:
```diff
@@ -326,3 +341,57 @@ definitions:
         example:
           message: "Recipient accepted the share"
           sharedSecret: "hfiuhworzwnur98d3wjiwhr"
+  AcceptedInvite:
```
Where is this used?

glpatcern replied:
Interesting, this is a left over from the previous PR. I'll drop it.

michielbdejong commented on spec.yaml:
```diff
@@ -139,16 +141,37 @@ paths:
+      summary: Inform the sender that the invitation was accepted to start sharing.
+      description: Inform about an accepted invitation so the user on the sender provider's side can initiate the OCM share creation.
+      parameters:
+        - name: invite
```
IIUC you're proposing two deviations from the current implementation as it exists in Reva:
1. change the endpoint from e.g. `https://sender.com/iop/ocm/invite/accept` to e.g. `https://sender.com/iop/ocm/invite-accepted` -> as agreed, that makes it clearer
2. change the JSON from:
```
{
  "token": "10431cd1-11df-49ce-af87-9dfc",
  "recipientProvider": "surf-iop-rd-app-acc.data.surfsara.nl",
  "userID": "marie",
  "name": "Marie Curie",
  "email":  "marie@surfsara.nl"
}
```
to (IIUC):
```
{
  "invite" : {
    "token": "10431cd1-11df-49ce-af87-9dfc",
    "recipientProvider": "surf-iop-rd-app-acc.data.surfsara.nl",
    "userID": "marie",
    "name": "Marie Curie",
    "email":  "marie@surfsara.nl"
  }
}
```
Can this second deviation be avoided? Does it serve a purpose?

glpatcern replied:
This was also not intended and comes from the original PR. Will fix it.

Edit: after deeper analysis it's good as is - thanks Michiel for double checking!

michielbdejong replied:
As discussed on Gitter, see https://swagger.io/docs/specification/2-0/describing-request-body/ - 
> The payload name. It is required but ignored (it is used for documentation purposes only).

So the existing current format for `/shares` is (confusing but) correct, and so is yours here!

michielbdejong replied:
Sorry for the confusion, you don't need to change anything.

glpatcern committed:
Removed spurious section

michielbdejong reviewed (approved):
I see you removed the spurious section and the 'name' in OpenAPI 2.0 does *not* add a level of nesting, so all good!
